replica/compaction_group.hh

```diff
@@ -233,23 +234,25 @@ protected:
     // The list entries are unlinked automatically when the storage group, they belong to, is removed.
     compaction_group_list _compaction_groups;
     storage_group_map _storage_groups;
-
+    // Prevents _storage_groups from having its elements inserted or deleted while other layer iterates
+    // over them (or over _compaction_groups).
+    seastar::rwlock _lock;
 public:
     virtual ~storage_group_manager();
 
+    seastar::rwlock& get_rwlock() noexcept {
+        return _lock;
+    }
+
     const compaction_group_list& compaction_groups() const noexcept {
         return _compaction_groups;
     }
     compaction_group_list& compaction_groups() noexcept {
         return _compaction_groups;
     }
 
-    const storage_group_map& storage_groups() const noexcept {
-        return _storage_groups;
-    }
-    storage_group_map& storage_groups() noexcept {
-        return _storage_groups;
-    }
+    future<> for_each_storage_group(const rwlock::holder&, std::function<future<>(size_t, storage_group&)> f) const;
```

Comment:
please add a comment explaining why the rwlock::holder is needed

Reply:
The lock holder is not used in for_each_storage_group at all

```
+future<> storage_group_manager::for_each_storage_group(const rwlock::holder&, std::function<future<>(size_t, storage_group&)> f) const {
+    for (auto& [id, sg]: _storage_groups) {
+        co_await f(id, *sg);
+    }
+}

```

It is enough to keep the lock holder at the call site.

Reply:
The reference to the lock holder is passed just as a proof that the caller holds a lock, otherwise it is easy to forget to acquire it.  I agree it's not very elegant and so it should be documented.

Reply:
The caller can pass a random holder to bypass it. It is better if we take the lock inside for_each_storage_group.